server/data_server.c

```diff
@@ -127,11 +130,22 @@ static void capture_experiment(void)
             count = hw_read_streamed_data(block, DATA_BLOCK_SIZE, &at_eof);
         while (data_thread_running  &&  count == 0  &&  !at_eof);
         if (count > 0)
+        {
+            if (!ts_captured)
+            {
+                hw_get_start_ts(&pcap_start_ts);
+                ts_captured = true;
+            }
             release_write_block(data_buffer, count);
+        }
 
         total_bytes += count;
         experiment_sample_count = total_bytes / sample_length;
     }
+
+    if (!ts_captured)
+        release_write_block(data_buffer, 0);
```

Comment:
Hmm.  On the face of it the condition `!ts_captured` should be replaced with `count == 0`, as each call to `get_write_block` ought to be balanced by a call to `release_write_block`, and this call will be omitted any time the loop above exits with `count` equal to 0.

On the other hand, `get_write_block` doesn't actually have any side effects, and calling `release_write_block` will trigger an empty update to the reader.  Still, I think I'd prefer a form that looks correct locally.